sljit_src/sljitNativeARM_64.c

```diff
@@ -2521,6 +2537,28 @@ SLJIT_API_FUNC_ATTRIBUTE sljit_s32 sljit_emit_atomic_store(struct sljit_compiler
 	CHECK_ERROR();
 	CHECK(check_sljit_emit_atomic_store(compiler, op, src_reg, mem_reg, temp_reg));
 
+#if __ARM_FEATURE_ATOMICS
+	switch (GET_OPCODE(op)) {
+	case SLJIT_MOV32:
+	case SLJIT_MOV_U32:
+		ins = CASL ^ (1 << 30);
+		break;
+	case SLJIT_MOV_U8:
+		ins = CASLB;
+		break;
+	case SLJIT_MOV_U16:
+		ins = CASLH;
+		break;
+	default:
+		ins = CASL;
+		break;
+	}
+	FAIL_IF(push_inst(compiler, atomic_load_ins(op) | RN(mem_reg) | RT(TMP_REG2)));
+	FAIL_IF(push_inst(compiler, SUBS | RM(TMP_REG2) | RN(temp_reg) | RD(TMP_REG1)));
+	FAIL_IF(push_inst(compiler, B_CC | (2 << 5) | 0x1));
+	FAIL_IF(push_inst(compiler, ins | RM(temp_reg) | RN(mem_reg) | RD(src_reg)));
+	FAIL_IF(push_inst(compiler, CLREX));
```

Comment:
This looks like a lot of instruction, Usually the point of these instruction is simplicity, why a single CAS instruction is not enough like on other cpus?

Reply:
It is still a RISC cpu, and having 3 arguments was probably alien enough that they forgot to return a flag to indicate if their CAS was successful or not.

if we are keeping the "undefined behaviour" if the temp register doesn't match then it could be done in a much simpler way, albeit IMHO not as robust:

  https://github.com/carenas/sljit/commit/b1e13c8064b45e7808833ceee35ed35f8e0a7885

If instead of doing only load/store, we are to provide atomic operations (like atomic add) then it will be easier to map them to the CPU instructions.

Reply:
Sljit has checks (asserts) to try to help when the api is misused, but in general, no runtime checks are generated, since they reduce the performance. If you want, you could try invent a way to check these load/store pairs at compile time, probably could help a lot.